java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java

```diff
@@ -181,26 +183,32 @@ public Boolean visit(ZeroVector left, Range range) {
   /**
    * Creates a visitor to visit child vectors.
    * It is used for complex vector types.
-   * @return the visitor for child vecors.
+   * @return the visitor for child vectors.
    */
   protected RangeEqualsVisitor createInnerVisitor(ValueVector leftInner, ValueVector rightInner) {
+    return this.createInnerVisitor(leftInner, rightInner, this.isTypeCheckNeeded);
+  }
+
+  protected RangeEqualsVisitor createInnerVisitor(
+          ValueVector leftInner, ValueVector rightInner, boolean isTypeCheckNeeded) {
     return new RangeEqualsVisitor(leftInner, rightInner, isTypeCheckNeeded);
   }
 
   protected boolean compareUnionVectors(Range range) {
     UnionVector leftVector = (UnionVector) left;
     UnionVector rightVector = (UnionVector) right;
 
-    List<FieldVector> leftChildren = leftVector.getChildrenFromFields();
-    List<FieldVector> rightChildren = rightVector.getChildrenFromFields();
-
-    if (leftChildren.size() != rightChildren.size()) {
-      return false;
-    }
+    Range subRange = new Range(0, 0, 1);
+    for (int i = 0; i < range.getLength(); i++) {
+      subRange.setLeftStart(range.getLeftStart() + i).setRightStart(range.getRightStart() + i);
+      ValueVector leftSubVector = leftVector.getVector(range.getLeftStart() + i);
+      ValueVector rightSubVector = rightVector.getVector(range.getRightStart() + i);
 
-    for (int k = 0; k < leftChildren.size(); k++) {
-      RangeEqualsVisitor visitor = createInnerVisitor(leftChildren.get(k), rightChildren.get(k));
-      if (!visitor.rangeEquals(range)) {
+      if (leftSubVector == null || rightSubVector == null) {
+        return leftSubVector == rightSubVector;
```

Comment:
shouldn't this only return on inequality?

Reply:
nice catch. Thank you so much.